### PR TITLE
Add feature to skip redirection on adding to cart

### LINKED ISSUE #689: No Add To Cart Redirect
Original idea came from [here](https://www.reddit.com/r/Steam/comments/e8hsjs/just_let_me_unwish_directly_from_the_storepage/fadd8md?utm_source=share&utm_medium=web2x).
Instead of automatically redirecting to cart after pressing "add to cart", it could show a dialog with an option to go to cart, to have an animation play and scroll to top.

### PR DESCRIPTION
**Pull request by candela97**
Closes #689.

This is a simple solution that sends a POST request to /cart/ and shows a "Continue Shopping" dialog on success. Works on app, bundle, sub pages, and wishlists.

### CODE REVIEW

**Comment by candela97** [comment]
oops, this is meant to be a draft

**Comment by Revadike** [comment]
Maybe scroll to top, instead of hard refreshing the page, would be better.

**Comment by candela97** [comment]
> I'd display a "Continue shopping?" style dialog when the user adds an item from the wishlist to the cart and doesn't have a value for the option set

@MxtOUT I think you meant for this PR?
OK, yeah constantly reloading on the wishlist seems unnecessary.

> Maybe scroll to top, instead of hard refreshing the page, would be better.
Hmm, on app pages the page will need to be refreshed to get the updated version, otherwise some stuff will be missing/invalid. We could adjust the HTML ourselves, but it seems like unnecessary work...

**Comment by MxtOUT** [comment]
> @MxtOUT I think you meant for this PR?

Ups, yes

**Comment by MxtOUT** [comment]
Can't we use the `webRequest` API `onBeforeRequest` event and intercept POST requests to https://store.steampowered.com/cart/ ?
Then we send this information to the content script that has registered a listener before.
To make sure we're only intercepting these form navigation requests, we could use:
- The request header `Sec-Fetch-Mode` (don't know if every UA will set this though)
- [`chrome.webRequest.ResourceType`](https://developer.chrome.com/docs/extensions/reference/webRequest/#type-ResourceType)
- `details.requestBody.formData` in the `onBeforeRequest` callback

**Comment by Revadike** [comment]
Why such a complicated solution when you can simply adapt `addToCart()` function?

**Comment by candela97** [comment]
> Why such a complicated solution when you can simply adapt `addToCart()` function?

Sadly I'm having less and less confidence in this "simple solution": there's another edge case with free promotions - those still show an add-to-cart button but redirects back to the store page afterwards, and it won't work on the new sale pages built with React.

Intercepting requests seems like neat idea, but it'll take me a while to learn how to use the `webRequest` API... If the request is intercepted, can we still verify if the item was added to cart successfully? There're some cases where it redirects to a different page, as I noted in the OP.

**Comment by MxtOUT** [comment]
> If the request is intercepted, can we still verify if the item was added to cart successfully?

You should be able to intercept the HTTP status code and the redirect URL (if present)

**Comment by Revadike** [comment]
> Sadly I'm having less and less confidence in this "simple solution": there's another edge case with free promotions - those still show an add-to-cart button but redirects back to the store page afterwards, and it won't work on the new sale pages built with React.

Just perform the request as an ajax/xhr request and check the finalUrl. Easy solution. You have to perform the POST request anyway.
Additionally, you can check the response HTML of this POST request (if redirect to cart) and use that to show the games in cart in the cart onhover event.
Do I make sense to you?

File: src/js/Content/Features/Store/Common/FAddToCartNoRedirect.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            if (!node) { return; }
+
+            const form = getFormEl(node.href);
+            if (!form) { return; }
+
+            e.preventDefault();
+
+            try {
+                await RequestData.post("https://store.steampowered.com/cart/", new FormData(form));
+            } catch (err) {} // ignore error
```
Comment (Revadike) [inline src/js/Content/Features/Store/Common/FAddToCartNoRedirect.js:49]: TODO: Update cart icon here (maybe with animation), with an onhover event that shows a mini-cart. Just like this example on greenmangaming:
image

**Comment by Revadike** [comment]
Also, you should consider these: https://store.steampowered.com/itemstore/252490/detail/43107/

**Comment by candela97** [comment]
Itemstores are already "dynamic" in a way that if you buy stuff a dialog will popup asking if you want to continue shopping or proceed to checkout, so it doesn't need this feature imo.

> Additionally, you can check the response HTML of this POST request (if redirect to cart) and use that to show the games in cart in the cart onhover event.

This should be done as a separate feature that fetches the user's cart page when they're on an app page.

**Comment by candela97** [comment]
This is now pretty much done and ready for testing. Changes:

1. Split out the code for sending the form through `RequestData.post` to a new module `AddToCart` (fell free to suggest better naming) so it can be reused in DLC checkboxes.
2. Display a "Continue Shopping?" dialog for both store pages and wishlists.

TODO Move strings to Localization, add option

File: src/css/augmentedsteam.css [hunk 2]
```diff
@@ -1358,10 +1358,6 @@ img.astats_icon {
   overflow-x: initial !important;
   overflow: visible !important;
 }
-.es_regional_prices.game_area_purchase_game:hover,
-.es_regional_prices.sale_page_purchase_item:hover {
-  z-index: 9999;
-}
```
Comment (candela97) [inline src/css/augmentedsteam.css:1364]: I'm not clear what these styles are for (added in https://github.com/jshackles/Enhanced_Steam/commit/e96b382abe9017a2a958d54502ffa7124c5f3d7e) but they glitch out the dialog so I removed them.

**Comment by candela97** [comment]
This is now pretty much done and ready for review. Changes:

1. Split out the code for sending the form through RequestData.post to a new module AddToCart (fell free to suggest better naming) so it can be reused for DLC checkboxes.
2. Display a "Continue Shopping?" dialog for both store pages and wishlists.

**Comment by Revadike** [comment]
> 2\. Display a "Continue Shopping?" dialog for both store pages and wishlists

Could you consider something maybe less intrusive? Perhaps how greenmangaming does it.

**Comment by candela97** [comment]
> > 2. Display a "Continue Shopping?" dialog for both store pages and wishlists
> 
> Could you consider something maybe less intrusive? Perhaps how greenmangaming does it.

I honestly think that's too complicated. Turning the page into some kind of pseudo-static page is definitely doable, but can easily break as Steam makes changes here and there. Simply reloading the page is an acceptable solution for now imo. As of now this works the way I envisioned it should, and I've spent too much time on this "simple solution" that might not get accepted already 😂.

**Comment by Revadike** [comment]
> can easily break as Steam makes changes here and there

I mean, that applies to everything.

**Comment by MxtOUT** [comment]
If you already give the user an option, this should just be "Prevent redirect to cart when adding items to cart" instead of showing yet another dialog.
So I'd just remove the dialog part of this feature, what do you think?

**Comment by candela97** [comment]
> If you already give the user an option, this should just be "Prevent redirect to cart when adding items to cart" instead of showing yet another dialog.
> So I'd just remove the dialog part of this feature, what do you think?

But I thought a dialog is what you wanted...
> I'd display a "Continue shopping?" style dialog when the user adds an item from the wishlist to the cart and doesn't have > a value for the option set

I just added it on apps page too to keep things consistent, otherwise there'll be another condition to add.

**Comment by MxtOUT** [comment]
> But I thought a dialog is what you wanted...

Oh yeah, I still do, but only for the first time a user could potentially use this feature, e.g. when the update rolls out and the user puts an item into his cart.
So we tell him: "Look at our neat feature, do you want to stay on the page for the future or keep it as it is?"
Then we have a decision and we don't need any dialogue interaction from the user when applying the feature

**Comment by Revadike** [comment]
> > But I thought a dialog is what you wanted...
> 
> Oh yeah, I still do, but only for the first time a user could potentially use this feature, e.g. when the update rolls out and the user puts an item into his cart.
> So we tell him: "Look at our new neat feature, do you want to stay on the page for the future or keep it as it is?"
> Then we have a decision and we don't need any dialogue interaction from the user when applying the feature

I think such dialog can be applied to many other features too.

**Comment by candela97** [comment]
> > But I thought a dialog is what you wanted...
> 
> Oh yeah, I still do, but only for the first time a user could potentially use this feature, e.g. when the update rolls out and the user puts an item into his cart.
> So we tell him: "Look at our new neat feature, do you want to stay on the page for the future or keep it as it is?"
> Then we have a decision and we don't need any dialogue interaction from the user when applying the feature

I worry default-enabled features like this (at least the dialog part) might put off some people, especially since purchasing is a core part of Steam. This doesn't solve having to add an extra option either since it's still needed in case some people want the old behaviour back.

How about adding a "Remember my choice" checkbox in the continue dialog? This way people who have explicitly enabled this feature can choose to dismiss the nagging dialog if they want.

**Comment by Revadike** [comment]
I think @MxtOUT would like that idea. He has suggested that for other stuff in the past.

**Comment by MxtOUT** [comment]
I'd still go for prompting first-time users (you can do this via `SyncedStorage.has("skipAddToCart")`) whether they want to have this feature enabled.
I don't think it is an intrusive way to introduce a feature.

And instead of giving you a dialogue, it will just apply the actual feature. Though I have read that a ["Continue shopping" button actually does something else](https://ux.stackexchange.com/search?q=continue+shopping) than what we're trying to achieve.

**Comment by MxtOUT** [comment]
So to summarize (I might have not been clear):
When the user is adding an item to his cart and the feature is active for the first time, a dialog opens and asks whether he wants to skip redirection in the future.
Now we're all set and we don't need a decision from the user every time he clicks on the "Add to cart" button.
This decision can always be changed on the options page.

**Comment by candela97** [comment]
For wishlists, since we don't want to refresh the page after adding an item, I think we can instead add a UI element to indicate the item has been added, along with a handy link to the shopping cart.

**Comment by candela97** [comment]
@MxtOUT  I think we still need the dialog that signals the item has been added to cart (esp. on wishlists, since we don't reload the page). Other than that, https://github.com/IsThereAnyDeal/AugmentedSteam/pull/1027#issuecomment-848913014 might be a better long-term solution after all, since this won't work on the new sale/category pages.

File: src/css/augmentedsteam.css (hunk 2 above)
Comment (candela97) [reply src/css/augmentedsteam.css:1330]: Turns out this is needed for the regional price popup to show on top of subsequent purchase containers. I think lowering the z-index will fix the dialog glitch.

**Pull request merged**